### ISSUE TITLE
Add bounce rate

### ISSUE THREAD

**Issue by adriaanvanrossum** [simpleanalytics/roadmap#16]
It would be based on the original page views feature.

**Comment by adriaanvanrossum** [comment]
Just thought about this a bit more. For bounce rate you need to know how many people did visit one page only. For these analytics:

- Visitor A visits page /page1
- Visitor B visits page /page1
- Visitor A visits page /page2

the bounce rate should be 1 / 3 = 33.33%.

> Bounce rate (%) = Visits that access only a single page (#) ÷ Total visits (#) to the website. [[1](https://en.wikipedia.org/wiki/Bounce_rate)]

We can measure if a user only visited one page by checking their referrer:

| Event | Learning | Unique page views for /page1 |
|-------|----------|-------|
| Visitor A visits page /page1 | No referrer = unique | 0 + 1 = 1 |
| Visitor B visits page /page1 | No referrer = unique | 1 + 1 = 2 |
| Visitor A visits page /page2 | Referrer = /page1 | 2 - 1 = 1 |

This way we can calculate the bounce rate per page.

**Comment by Jivings** [comment]
Is this really how it is calculated? I always assumed it to be based on visits/visitors.

eg if there are two visitors and one bounces then I would expect bounce rate to be 50%. Because one out of two visitors bounced.

**Comment by adriaanvanrossum** [comment]
So total visits should be unique visitors? That makes sense. Will update it above.

**Comment by mtimofiiv** [comment]
One thing I guess I can think of that slightly complicates things.

 * Visitor A visits page `/page1`
 * Visitor B visits page `/page1`
 * Visitor A visits page `/page2`
 * Visitor A visits page `/page3`

The bounce rate here should still be 1 / 2 = 50% still.

So if you're solely checking the referrer to determine `- 1`, you will get an inaccurate bounce rate since in that case you will have 0 / 2 = 0%. No one ever leaves 😉

**Comment by adriaanvanrossum** [comment]
Just a bit more thinking on the bounce rate without cookies or sessions. What happens if you record the last path of every page?

Let’s say you have 3 pages:
- /
- /product
- /contact

You have 3 visitors that take this route:
1. / > /contact
2. / > /product
3. /contact

Total bounce rate should be 33%. Visitor 3 is not going to another page. But how do you know without a session?

You always know the entry page (no referrer or a referrer of a different website). So the entry pages here are:
- 2x /
- 1x /contact

Because you record the last page for every page visited you can get all pages that have entry pages as their last page.

To get the bounce rate you calculate: 
Entry pages (3) - pages that have entry pages as their last page (2) / entry pages (3) = 33% 

This seems to work until you get this:
1. / > /contact > /product
2. / > /product
3. /contact

Bounce rate = Entry pages (3) - pages that have entry pages as their last page (3) / entry pages (3) = 100%

Hm

**Comment by Jivings** [comment]
I thought you were going to use referrer header for this?

If user goes from `james.com/` to `james.com/contact` then the referrer will be `james.com` so you know it wasn't their entry page.

If user lands on `/contact` then bounces then their referrer will be some external site, so you know it was their entry page.

**Comment by adriaanvanrossum** [comment]
Yeah, last path or referrer would be the same. Calculating the entry pages are not the problem here. Calculating the click through is the problem. But you can't really when you have this:

1. `/` > `/contact` > `/product`
2. `/` > `/product`
3. `/contact`

Because how would you know it wasn't this:

1. `/` > `/contact`
2. `/` > `/product`
3. `/contact` > `/product`

**Comment by adriaanvanrossum** [comment]
Maybe we should add bounce rate for SPA's and move people to a SPA when they want a bounce rate.

**Comment by Jivings** [comment]
> Yeah, last path or referrer would be the same. Calculating the entry pages are not the problem here. Calculating the click through is the problem. 

Oh right, you mean you don't know if a user has bounced or not because you aren't tracking the relationship between the page they landed on and any subsequent page they visited. Yes I also can't see a way to do this without cookies on a non-SPA...
